library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -232,48 +178,46 @@ impl<T, A: Allocator> VecDeque<T, A> {
     /// Returns `true` if the buffer is at full capacity.
     #[inline]
     fn is_full(&self) -> bool {
-        self.cap() - self.len() == 1
+        self.len == self.capacity()
     }
 
     /// Returns the index in the underlying buffer for a given logical element
-    /// index.
+    /// index + addend.
     #[inline]
-    fn wrap_index(&self, idx: usize) -> usize {
-        wrap_index(idx, self.cap())
+    fn wrap_add(&self, idx: usize, addend: usize) -> usize {
+        wrap_index(idx.wrapping_add(addend), self.capacity())
```

Comment:
pondering: this should *never* actually wrap, right?  The addend should be such that `idx + addend < capacity * 2`, and if it *does* wrap then it's going to give the wrong element?

If I'm correct about that, I think this should be
```suggestion
        wrap_index(idx + addend, self.capacity())
```
so that the debug builders in the bors tests have the opportunity to catch such a problem, if it ever gets made, while still doing the same thing in release mode as the `wrapping_add`.

Reply:
If `T::IS_ZST`, then `head` can go all the way up to `usize::MAX` (e.g. just by doing one `push_front`). In that case, the addition can in fact overflow. For all non-ZSTs, the addition should never overflow though.

Reply:
Hmm.  That makes me wonder if LLVM could do better knowing that this doesn't actually overflow, but telling it that probably needs lots of `unsafe` so definitely shouldn't be this PR.